### PR TITLE
feat(ui5-tokenizer): truncate when having 1 token with long text

### PR DESCRIPTION
**Pull request by MapTo0**

### CODE REVIEW

**Comment by MapTo0** [comment]
FIXES: #5229

File: packages/main/src/Tokenizer.ts [hunk 1]
```diff
@@ -251,6 +276,10 @@ class Tokenizer extends UI5Element {
 		const target = e.target as Token;
 		if (!e.detail) { // if there are no details, the event is triggered by a click
```
Comment (niyap) [inline packages/main/src/Tokenizer.ts:277]: If you have one long token and another one, so the long token is expanded. Trying to delete the long token from its clear icon it is not possible, it is just scrolled to the beginning, _delete is not invoked at all. In order to be able to delete it from its clear icon, it should be the only token in the ui5-multi-input/combobox.

File: packages/main/src/Tokenizer.ts [hunk 2]
```diff
@@ -229,8 +248,14 @@ class Tokenizer extends UI5Element {
 		});
 	}
 
+	async popoverListFocusOut() {
```
Comment (niyap) [inline packages/main/src/Tokenizer.ts:251]: When you have one long token and its popover is open, if you delete the token from the popover, for a second an empty popover is shown and then the popover is closed at all. Can we prevent that?

File: packages/main/src/Token.ts [hunk 3]
```diff
@@ -226,6 +232,14 @@ class Token extends UI5Element implements ITabbable {
 		return "decline";
 	}
 
+	get textDom() {
+		return this.getDomRef()!.querySelector(".ui5-token--text");
+	}
+
+	get isTruncatable() {
```
Comment (niyap) [inline packages/main/src/Token.ts:239]: Is it expected when we have single long token to be truncated and not fully visible when the focus is in the input field and when we have more that one token and the focus is in the input field, then the long one is expanded?

**Review by niyap — CHANGES_REQUESTED** [review]
1. When the focus is in the suggestions/n-more list and we have one long token and other tokens/suggestions, Shift + Tab returns the focus on the token. In the other scenario, the focus goes to the previous focusable element on the page
2. In the MultiComboBox, when the focus is in the input field, if we have long token/ even long suggestion without token created and a token presented within the input field, on Shift + Tab the focus goes over the token. In the other use cases, it goes to the previous focusable element on the page.
3. If the control is readonly, again we have long token, click over the N-more link, then close the popover with Escape -> the focus is returned to the input field, then Shift + Tab, the focus goes again over the token instead of going to the previous focusable element on the page.
4. When we have multiple tokens, so the long one is expanded, move the focus over it and try to scroll it via mouse, on mouse up and the mouse down again, you do not continue scrolling but start from the beginning again. Reproducible in both controls, you can compare the behaviour with UI5
5. If we have one long token and open its popover, on Shift + Tab, the focus is lost (moved to the body) instead of moved to the previous element
6. Dialog representing the long token's popover on mobile looks really strange. Is that the expected design?
7. The suggestions popover has max-width set to 40rem, where the long token's popover takes the whole screen available width. Isn't it expected to be restricted to 40rem as well and then when the wrapping is enabled to takes more then one line in such cases?

File: packages/main/test/pages/MultiInput.html [hunk 4]
```diff
@@ -226,8 +226,15 @@ <h1 class="sample-container-title">Multi Input with 5 tokens</h1>
 		<br>
 		<br>
 
-
 		<ui5-button id="add-to-multiple">Add more tokens</ui5-button>
+
+		<br>
```
[the remark this answers is not in the document]
Comment (MapTo0) [reply packages/main/test/pages/MultiInput.html:231]: There is already such example check `#token-unique` element. I have added a test to check that behaviour.

File: packages/main/src/themes/Tokenizer.css [hunk 5]
```diff
@@ -44,6 +44,13 @@
 	padding-inline-start: var(--_ui5_tokenizer_padding);
 }
 
+:host([_tokens-count="1"]) .ui5-tokenizer--content {
+	padding-inline-end: 4px;
```
[the remark this answers is not in the document]
Comment (MapTo0) [reply packages/main/src/themes/Tokenizer.css:48]: the container has an overflow hidden set but the focus of the token is set to be outside and it overflows its content so it gets cut.
image

File: packages/main/src/Tokenizer.ts (hunk 2 above)
Comment (MapTo0) [reply packages/main/src/Tokenizer.ts:251]: cant reproduce that

**Comment by MapTo0** [comment]
> 1. When the focus is in the suggestions/n-more list and we have one long token and other tokens/suggestions, Shift + Tab returns the focus on the token. In the other scenario, the focus goes to the previous focusable element on the page

**Focus should stay on the list when hitting Tab / Shift + Tab as in UI5**
_Fixed that in latest changes_

> 2. In the MultiComboBox, when the focus is in the input field, if we have long token/ even long suggestion without token created and a token presented within the input field, on Shift + Tab the focus goes over the token. In the other use cases, it goes to the previous focusable element on the page.

**Focus should skip Tokenizer when on input field and tab / shift tab is used**
_This is another issue that is present before that implementation and I would fix it in another change_

> 3. If the control is readonly, again we have long token, click over the N-more link, then close the popover with Escape -> the focus is returned to the input field, then Shift + Tab, the focus goes again over the token instead of going to the previous focusable element on the page.

**Focus should skip Tokenizer when Multi Input / Multi Combo Box is readonly**
_This is another issue that is present before that implementation and I would fix it in another change_


> 4. When we have multiple tokens, so the long one is expanded, move the focus over it and try to scroll it via mouse, on mouse up and the mouse down again, you do not continue scrolling but start from the beginning again. Reproducible in both controls, you can compare the behaviour with UI5

_Fixed that in latest changes_

> 5. If we have one long token and open its popover, on Shift + Tab, the focus is lost (moved to the body) instead of moved to the previous element

_Fixed that in latest changes_

> 6. Dialog representing the long token's popover on mobile looks really strange. Is that the expected design?

**I aligned a bit the behaviour with UI5's Dialog but this used to be like that before my change**

> 7. The suggestions popover has max-width set to 40rem, where the long token's popover takes the whole screen available width. Isn't it expected to be restricted to 40rem as well and then when the wrapping is enabled to takes more then one line in such cases?

**That will be implemented within separate BLI as discussed with PO**

File: packages/main/src/themes/TokenizerPopover.css [hunk 6]
```diff
@@ -2,3 +2,8 @@
 [ui5-responsive-popover]::part(content) {
 	padding: 0;
 }
+
+#ui5-tokenizer-dialog-confirm-btn {
```
Comment (niyap) [inline packages/main/src/themes/TokenizerPopover.css:6]: isn't is better to use classes

**Review by niyap — CHANGES_REQUESTED** [review]
1. Now, when we open the long token's dropdown and the focus is there on Tab or Shift+Tab, the focus does not move, it is trapped there. I noticed that in UI5 it behaves in the same way but I do not think that it is expected, as a blind user is not able to navigate once within the popover.
2. If we open the long token's popover and click the arrow down of the MultiComboBox in order to open the suggestions popover, both popovers are visible.
3. As discussed, please check once again the problem with the empty popover, when you try to delete the long token from its popover.
4. I agree to follow the UI5 approach for the dialog currently, but please inform the responsible designers that we need a specification for the dialog of long token on mobile
5. Please, check the #token-unique sample in compact mode, create a long token and open its popover -> it overlaps the input field and looks strange. It is also reproducible in some other scenario. I know that it is not related to that change, but please open an issue.
6. Now, if we have ui5-multi-input with N-more and we select via click one of the visible (not overflowed tokens) then trying to move the focus to the previous visible token with Arrow left, it is not possible. You can try #multiple-token, select "eu" token and try to move with arrow left to "ea" token.
7. Please create an issue for the problem with the ItemNavigation and token focus as well.

**Comment by MapTo0** [comment]
> Now, when we open the long token's dropdown and the focus is there on Tab or Shift+Tab, the focus does not move, it is trapped there. I noticed that in UI5 it behaves in the same way but I do not think that it is expected, as a blind user is not able to navigate once within the popover.

**As stated, this is the behaviour in openui5. Furthermore, we have special logic in all popups to keep the focus inside and navigate back to the page only when the current open popup is closed. Therefore, I would like to leave this as it is.**

> If we open the long token's popover and click the arrow down of the MultiComboBox in order to open the suggestions popover, both popovers are visible.

**Fixed**

> As discussed, please check once again the problem with the empty popover, when you try to delete the long token from its popover.

**Fixed**

> I agree to follow the UI5 approach for the dialog currently, but please inform the responsible designers that we need a specification for the dialog of long token on mobile

**Done**

> Please, check the #token-unique sample in compact mode, create a long token and open its popover -> it overlaps the input field and looks strange. It is also reproducible in some other scenario. I know that it is not related to that change, but please open an issue.

**Done**

> Now, if we have ui5-multi-input with N-more and we select via click one of the visible (not overflowed tokens) then trying to move the focus to the previous visible token with Arrow left, it is not possible. You can try #multiple-token, select "eu" token and try to move with arrow left to "ea" token.

> Please create an issue for the problem with the ItemNavigation and token focus as well.

**This is an existing issue in master as well. https://sap.github.io/ui5-webcomponents/main/playground/main/pages/MultiInput/?sap-ui-theme=sap_horizon**
**Reported in #7256**

**Review by niyap — APPROVED** [review]

**Pull request merged**